modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/webpack_federation.js

```diff
@@ -12,61 +12,198 @@
-		else {
-			const script = document.createElement('script');
+		document.body.appendChild(script);
+
+		return containerRequest;
+	}
+
+	/**
+	 * Log messages if `explain resolutions` option is turned on.
+	 */
+	function explain(...things) {
+		// Until we get to the end of webpack 5 migration we will log all
+		// messages no matter what is configured. Once finished, we will only
+		// write when the old existing `explain resolutions` option is on.
+		console.log(...things);
+	}
 
-			script.src = '/o/' + containerId + '/__generated__/container.js';
+	/**
+	 * Fetch a container's module by path
+	 */
+	function fetch(containerId, path, resolve, reject) {
+		const url = '/o/' + containerId + '/__generated__/container.js';
 
-			script.onload = () => {
-				container = Liferay.Webpack.Container[containerId];
+		explain('Fetching container', containerId, 'from', url);
+
+		containerRequests[containerId] = createContainerRequest(
+			url,
+			(containerRequest) => {
+				explain('Fetched container', containerId);
+
+				const container = getContainer(containerId);
 
 				if (container) {
-					Promise.resolve(container.init(Liferay.Webpack.SharedScope))
-						.then(() => container.get(path))
-						.then((module) => resolve(module()));
+					explain(
+						'Initializing container', containerId, '\n',
+						Object.entries(sharedScope).map(([name, data]) =>
+							`${name}: ` +
+							Object.entries(data).map(
+								([version, data]) => `(${version}: ${data.from})`
+                            )
+						).join('\n ')
+					);
+
+					Promise.resolve(
+						container.init(sharedScope)
+					)
+					.then(
+						() => container.get(path)
+					)
+					.then(
+						(moduleFactory) => {
+							finalizeContainerRequest(
+								containerRequest, moduleFactory());
+						}
+					);
 				}
 				else {
 					const message =
-						'Container ' +
-						containerId +
-						' was fetched but ' +
-						'did not define itself in Liferay.Webpack.Container';
+						`Container ${containerId} was fetched but its script ` +
+						`failed to register with Liferay`;
 
 					console.warn(message);
 
-					reject(new Error(message));
+					finalizeContainerRequest(
+						containerRequest, new Error(message));
 				}
-			};
+			}
+		);
 
-			document.body.appendChild(script);
+
+		subscribeContainerRequest(
+			containerRequests[containerId], resolve, reject);
+	}
+
+	/**
+	 * Finalize an ongoing container request: set its result (error or module)
+	 * and notify all pending subscribers.
+	 */
+	function finalizeContainerRequest(containerRequest, result) {
+		const {script, subscribers} = containerRequest;
+
+		containerRequest.fetched = true;
+		containerRequest.script = undefined;
+		containerRequest.subscribers = undefined;
+
+		if(result instanceof Error) {
+			explain('Rejecting container', script.src, 'for', subscribers.length, 'subscribers');
+			containerRequest.error = result;
+			subscribers.forEach(({reject}) => reject(result));
+		} else {
```

Comment:
I'll shut up about SF now...

Reply:
yes, because I didn't pass the SF at all (being draft and unfinished :-( )